Sources/Leaf/Renderer/Renderer.swift

```diff
@@ -0,0 +1,69 @@
+import Bits
+
+/// Renders Leaf templates using the Leaf parser and serializer.
+public final class Renderer {
+    /// The tags available to this renderer.
+    public let tags: [String: Tag]
+
+    /// The renderer will use this to read files for
+    /// tags that require it (such as #embed)
+    public let fileReader: FileReader
+
+    /// Create a new Leaf renderer.
+    public init(tags: [String: Tag]? = nil, fileReader: FileReader) {
+        self.tags = tags ?? defaultTags
+        self.fileReader = fileReader
+    }
+
+    // ASTs only need to be parsed once
+    private var _cachedASTs: [Int: [Syntax]] = [:]
+
+    /// Renders the supplied template bytes into a view
+    /// using the supplied context.
+    public func render(_ template: Bytes, context: DataRepresentable) throws -> Bytes {
+        let hash = template.makeString().hashValue
+
+        let ast: [Syntax]
+        if let cached = _cachedASTs[hash] {
+            ast = cached
+        } else {
+            let parser = Parser(template)
+            do {
+                ast = try parser.parse()
+            } catch let error as ParserError {
+                throw RenderError(source: error.source, reason: error.reason, error: error)
+            }
+            _cachedASTs[hash] = ast
```

Comment:
Is there a reason to store the hash instead of storing the string and getting the proper handling of hash conflicts. I think right now when we have a hash conflict it's going to make some really weird bugs for people.

Probably pretty rare since templates are generally larger, but probably better to be safe than sorry.